nemo_aligner/models/mm/stable_diffusion/megatron_sd_draftp_model.py

```diff
@@ -0,0 +1,468 @@
+                rewards = self.reward_model.get_reward(output_tensor_draft_p.permute(0, 2, 3, 1), dataloader_iter)
+                loss = -rewards.mean() + kl_penalty * self.cfg.kl_coeff
+
+                reduced_loss = average_losses_across_data_parallel_group([loss])
+
+                return (
+                    loss,
+                    {"loss": reduced_loss, "kl_penalty": kl_penalty},
+                )
+
+            return output_tensor_draft_p, loss_func
+
+        return fwd_output_and_loss_func
+
+    def calculate_gaussian_kl_penalty_shared_var(self, curr_eps, init_eps):
+
+        diff = curr_eps - init_eps
+        kl = torch.sum(diff ** 2, dim=(1, 2, 3, 4), keepdim=True).flatten()
+        dimensionality = torch.numel(curr_eps[0])
+        kl /= dimensionality
+
+        return kl
+
+    def get_loss_and_metrics(self, batch, forward_only=False):
+
+        fwd_bwd_function = get_forward_backward_func()
+
+        losses_reduced_per_micro_batch = []
+
+        # TODO @ataghibakhsh: input num_microbatch manually
+        for i in range(0, len(batch), self.cfg.micro_batch_size):
+            data_item = batch[i : i + self.cfg.micro_batch_size]
+
+            losses_reduced_per_micro_batch.append(
+                fwd_bwd_function(
+                    forward_step_func=self.get_forward_output_and_loss_func(),
+                    data_iterator=[data_item],  # 4
+                    model=self.model.model,
+                    num_microbatches=self.cfg.micro_batch_size,  # 4
+                    forward_only=forward_only,
+                    seq_length=None,
+                    micro_batch_size=self.cfg.micro_batch_size,  # 1
+                )[0]
+            )
+
+        if torch.distributed.get_rank() == 0 and len(self.logger.loggers) > 1:
+            self.log_visualization(batch[0:1])
+
+        if self.model.with_distributed_adam:
+            # synchronize asynchronous grad reductions
+            # note: not necessary, but reduces performance degradation
+            # from multiple simultaneous NCCL calls
+            self._optimizer._finish_bucket_grad_sync()
+        else:
+            # async grad allreduce is not currently implemented for O1/autocasting mixed precision training
+            # so we all-reduce gradients after the pipeline
+            self.model.allreduce_gradients()
+
+        metrics = {}
+
+        for key in ["loss", "kl_penalty"]:
+            if losses_reduced_per_micro_batch:
+                metric_mean = torch.stack(
+                    [loss_reduced[key] for loss_reduced in losses_reduced_per_micro_batch]
+                ).mean()
+            else:
+                metric_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+
+            torch.distributed.broadcast(metric_mean, get_last_rank())
+
+            metrics[key] = metric_mean.cpu().item()
+
+        return metrics["loss"], metrics
+
+    def grad_reductions(self):
```

Comment:
we have a grad reduction function in [train_utils.py](https://github.com/NVIDIA/NeMo-Aligner/blob/main/nemo_aligner/utils/train_utils.py#L84)

Reply:
Done